packages/authorization-process/src/services/readModelService.ts

```diff
@@ -46,6 +59,89 @@ export function readModelServiceBuilder(
     ): Promise<WithMetadata<Client> | undefined> {
       return getClient({ "data.id": id });
     },
+
+    async getClients(
+      filters: GetClientsFilters,
+      { offset, limit }: { offset: number; limit: number }
+    ): Promise<ListResult<Client>> {
+      const { name, userIds, consumerId, purposeId, kind } = filters;
+
+      const nameFilter: ReadModelFilter<Client> = name
+        ? {
+            "data.name": {
+              $regex: ReadModelRepository.escapeRegExp(name),
+              $options: "i",
+            },
+          }
+        : {};
+
+      const userIdsFilter: ReadModelFilter<Client> =
+        ReadModelRepository.arrayToFilter(userIds, {
+          $or: userIds.map((userId) => ({ "data.users": { $eq: userId } })),
+        });
+
+      const consumerIdFilter: ReadModelFilter<Client> = {
+        "data.consumerId": { $eq: consumerId },
+      };
+
+      const purposeIdFilter: ReadModelFilter<Client> = purposeId
+        ? {
+            "data.purposes": { $eq: purposeId },
+          }
+        : {};
+
+      const kindFilter: ReadModelFilter<Client> = kind
+        ? {
+            "data.kind": { $eq: kind },
+          }
+        : {};
+
+      const aggregationPipeline = [
+        {
+          $match: {
+            ...nameFilter,
+            ...userIdsFilter,
+            ...consumerIdFilter,
+            ...purposeIdFilter,
+            ...kindFilter,
+          } satisfies ReadModelFilter<Client>,
+        },
+        {
+          $project: {
+            data: 1,
+            computedColumn: { $toLower: ["$data.name"] },
+          },
+        },
+        {
+          $sort: { computedColumn: 1 },
+        },
+      ];
+
+      const data = await clients
+        .aggregate(
+          [...aggregationPipeline, { $skip: offset }, { $limit: limit }],
+          { allowDiskUse: true }
+        )
+        .toArray();
+
+      const result = z.array(Client).safeParse(data.map((d) => d.data));
+      if (!result.success) {
+        throw genericInternalError(
+          `Unable to parse client items: result ${JSON.stringify(
+            result
+          )} - data ${JSON.stringify(data)} `
+        );
+      }
+
+      return {
+        results: result.data,
+        totalCount: await ReadModelRepository.getTotalCount(
+          clients,
+          aggregationPipeline,
```

Comment:
in this implementation `totalCount` calculation includes `sort` step.
Sorting is a heavy operation and is not required in this case.
I suggest to include it only in the `result` pipeline